### PR TITLE
feat: add Qwen3-Embedding-0.6B and Qwen3-Reranker-0.6B support

### LINKED ISSUE #528: [Model]: `Qwen/Qwen3-Embedding-0.6B-GGUF`
### Which model would you like to support?

Hi,

Would it be possible to add support for the model?
Link to the model: https://huggingface.co/Qwen/Qwen3-Embedding-0.6B-GGUF

Thank you!

### What are the main advantages of this model?

> The embedding model has achieved state-of-the-art performance across a wide range of downstream application evaluations. The 8B size embedding model ranks No.1 in the MTEB multilingual leaderboard (as of May 26, 2025, score 70.58), while the reranking model excels in various text retrieval scenarios.

### LINKED ISSUE #529: [Feature]: Last token pooling for causal embedding models
### What feature would you like to request?

The Qwen3 models will need something like this (this is taken from Qwen3 example):

```python
def last_token_pool(last_hidden_states: Tensor,
                 attention_mask: Tensor) -> Tensor:
    left_padding = (attention_mask[:, -1].sum() == attention_mask.shape[0])
    if left_padding:
        return last_hidden_states[:, -1]
    else:
        sequence_lengths = attention_mask.sum(dim=1) - 1
        batch_size = last_hidden_states.shape[0]
        return last_hidden_states[torch.arange(batch_size, device=last_hidden_states.device), sequence_lengths]
```

### Is there any additional information you would like to provide?

_No response_

### PR DESCRIPTION
**Pull request by n24q02m**
## Summary

Add native support for **Qwen3-Embedding-0.6B** and **Qwen3-Reranker-0.6B** models, including both **INT8** and **Q4F16** quantization variants.

These are the first causal-LM-based embedding and reranker models in fastembed, using fundamentally different architectures from existing BERT-family models.

Closes #528
Closes #529
Related to #530

### Qwen3TextEmbedding (`fastembed/text/qwen3_embedding.py`)
- **Last-token pooling** instead of CLS/mean pooling (causal LM architecture)
- **Matryoshka Representation Learning (MRL)**: truncate embeddings to any dimension 32-1024 via `dim=` parameter
- **Instruction-aware**: queries use `Instruct: {task}\nQuery: {text}` format
- Static ONNX batch constraint: hardcoded `batch_size=1` (causal LM ONNX graph limitation)
- **Two variants**: INT8 (default) and Q4F16 (smaller, INT4 weights + FP16 activations)

### Qwen3CrossEncoder (`fastembed/rerank/cross_encoder/qwen3_cross_encoder.py`)
- **Causal LM yes/no logit scoring** instead of traditional relevance head
- **Chat-template formatting** with system/user/assistant turns
- Extracts last-token logits for "yes"/"no" tokens, applies softmax -> P(yes)
- One-at-a-time processing (static batch=1 ONNX constraint)
- **Two variants**: INT8 (default) and Q4F16

### Shared Changes
- `last_token_pool()` utility in `fastembed/common/utils.py`
- `LAST_TOKEN` pooling type added to `PoolingType` enum
- `preprocessor_utils.py`: graceful handling of missing `special_tokens_map.json` + fix for `pad_token_id: null` and dict `pad_token` in tokenizer configs
- `model_management.py`: fix snapshot cache verification -- verify requested `model_file` exists in cached snapshot before returning (prevents stale cache hit when multiple variants share the same HF repo)
- `onnx_text_model.py`: cast float16 outputs to float32 after inference (required for Q4F16 variant)

### Models

| Model Name | Variant | File | Size |
|------------|---------|------|------|
| `Qwen/Qwen3-Embedding-0.6B` | INT8 (default) | `onnx/model_quantized.onnx` | 573 MB |
| `Qwen/Qwen3-Embedding-0.6B-Q4F16` | Q4F16 | `onnx/model_q4f16.onnx` | 517 MB |
| `Qwen/Qwen3-Reranker-0.6B` | INT8 (default) | `onnx/model_quantized.onnx` | 573 MB |
| `Qwen/Qwen3-Reranker-0.6B-Q4F16` | Q4F16 | `onnx/model_q4f16.onnx` | 518 MB |

### Tests
- Canonical vector values for `Qwen/Qwen3-Embedding-0.6B` in `test_text_onnx_embeddings.py`
- Canonical score values for `Qwen/Qwen3-Reranker-0.6B` in `test_text_cross_encoder.py`
- All existing tests unaffected

## ONNX Models

Pre-converted ONNX models hosted on HuggingFace:
- [n24q02m/Qwen3-Embedding-0.6B-ONNX](https://huggingface.co/n24q02m/Qwen3-Embedding-0.6B-ONNX) -- INT8 + Q4F16 variants
- [n24q02m/Qwen3-Reranker-0.6B-ONNX](https://huggingface.co/n24q02m/Qwen3-Reranker-0.6B-ONNX) -- INT8 + Q4F16 variants

Conversion pipeline:
- **INT8**: `torch.onnx.export` (opset 21) + `onnxruntime.quantization.quantize_dynamic` (QInt8)
- **Q4F16**: `MatMulNBitsQuantizer` (4-bit, block_size=128, symmetric) + FP16 cast

## Usage

```python
from fastembed import TextEmbedding, TextCrossEncoder

# Embedding (INT8 default)
model = TextEmbedding("Qwen/Qwen3-Embedding-0.6B")
embeddings = list(model.embed(["Hello world"]))               # 1024-dim
embeddings_256 = list(model.embed(["Hello world"], dim=256))  # MRL

# Embedding (Q4F16 -- smaller model)
model_q4 = TextEmbedding("Qwen/Qwen3-Embedding-0.6B-Q4F16")

# Query with instruction
query_emb = list(model.query_embed("What is machine learning?"))

# Reranker
reranker = TextCrossEncoder("Qwen/Qwen3-Reranker-0.6B")
scores = list(reranker.rerank("What is AI?", ["AI is...", "Pizza is..."]))
```

## Verification

Canonical values verified against the ONNX model outputs:
- **Embedding**: `[-0.0223, 0.0187, -0.0145, -0.0854, 0.0122]` (first 5 dims, "hello world")
- **Reranker**: `[0.9945, 0.0164]` ("What is the capital of France?" vs Paris/Berlin docs)

INT8 and Q4F16 variants both produce correct semantic rankings.

## Checklist

- [x] Model implementation files (INT8 + Q4F16 variants)
- [x] Registry integration (TextEmbedding, TextCrossEncoder)
- [x] Utility functions (`last_token_pool`)
- [x] Canonical test values
- [x] Snapshot cache fix for multi-variant model downloads
- [x] Float16 output handling
- [x] `ruff check` passes
- [x] `ruff format` passes
- [x] `mypy` passes (strict mode)

### CODE REVIEW

**Comment by n24q02m** [comment]
Closing — I'm no longer maintaining this contribution. Feel free to use the implementation as reference if the team wants to add Qwen3 support.
